File: android/app/src/main/kotlin/com/marrakechguide/feature/homebase/HomeBaseViewModel.kt
package com.marrakechguide.feature.homebase

import android.location.Location
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.marrakechguide.core.engine.GeoEngine
import com.marrakechguide.core.repository.HomeBase
import com.marrakechguide.core.repository.UserSettingsRepository
import com.marrakechguide.core.service.HeadingConfidence
import com.marrakechguide.core.service.HeadingService
import com.marrakechguide.core.service.LocationError
import com.marrakechguide.core.service.LocationService
import com.marrakechguide.core.service.PermissionStatus
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.math.ceil

/**
 * ViewModel for the Home Base compass feature.
 *
 * Manages:
 * - Loading/saving home base from settings
 * - Combining location and heading data
 * - Computing compass arrow rotation
 * - Formatting distance and time estimates
 */
@HiltViewModel
class HomeBaseViewModel @Inject constructor(
    private val settingsRepository: UserSettingsRepository,
    private val locationService: LocationService,
    private val headingService: HeadingService
) : ViewModel() {

    // MARK: - State

    private val _homeBase = MutableStateFlow<HomeBase?>(null)
    val homeBase: StateFlow<HomeBase?> = _homeBase.asStateFlow()

    private val _isLoading = MutableStateFlow(true)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    val permissionStatus: StateFlow<PermissionStatus> = locationService.permissionStatus

    val hasLocationPermission: Boolean
        get() = locationService.isAuthorized

    val headingConfidence: StateFlow<HeadingConfidence> = headingService.headingConfidence

    // MARK: - Computed State

    /**
     * Arrow rotation angle for the compass.
     * Combines device heading with bearing to home base.
     */
    val arrowRotation: StateFlow<Double> = combine(
        locationService.currentLocation,
        headingService.currentHeading,
        _homeBase
    ) { location, heading, homeBase ->
        computeArrowRotation(location, heading, homeBase)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0.0)

    /**
     * Current distance to home base in meters.
     */
    val distanceMeters: StateFlow<Double?> = combine(
        locationService.currentLocation,
        _homeBase
    ) { location, homeBase ->
        computeDistance(location, homeBase)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), null)

    /**
     * Formatted distance string (e.g., "450 m" or "1.2 km").
     */
    val formattedDistance: StateFlow<String> = combine(
        distanceMeters
    ) { (distance) ->
        distance?.let { GeoEngine.formatDistance(it) } ?: "--"
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), "--")

    /**
     * Estimated walk time string (e.g., "10 min").
     */
    val estimatedWalkTime: StateFlow<String> = combine(
        distanceMeters,
        locationService.currentLocation
    ) { distance, location ->
        computeWalkTimeString(distance, location)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), "--")

    /**
     * Cardinal direction description (e.g., "Northeast").
     */
    val directionDescription: StateFlow<String> = combine(
        arrowRotation
    ) { (rotation) ->
        cardinalDirection(rotation)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), "")

    // MARK: - Lifecycle

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                _homeBase.value = settingsRepository.getHomeBase()
                locationService.checkPermission()
            } catch (e: Exception) {
                _errorMessage.value = "Failed to load home base: ${e.message}"
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun startTracking() {
        if (!locationService.isAuthorized) return

        locationService.startUpdates()
        headingService.startUpdates()
    }

    fun stopTracking() {
        locationService.stopUpdates()
        headingService.stopUpdates()
    }

    fun refreshLocation() {
        viewModelScope.launch {
            try {
                locationService.refreshLocation()
            } catch (e: LocationError) {
                _errorMessage.value = e.message
            } catch (e: Exception) {
                _errorMessage.value = "Failed to refresh location: ${e.message}"
            }
        }
    }

    fun clearError() {
        _errorMessage.value = null
    }

    override fun onCleared() {
        super.onCleared()
        stopTracking()
    }

    // MARK: - Private Methods

    private fun computeArrowRotation(
        location: Location?,
        heading: Float?,
        homeBase: HomeBase?
    ): Double {
        if (location == null || homeBase == null) return 0.0
        if (heading == null) return 0.0

        val currentCoord = GeoEngine.Coordinate(location.latitude, location.longitude)
        val homeCoord = GeoEngine.Coordinate(homeBase.lat, homeBase.lng)

        val bearingToHome = GeoEngine.bearing(currentCoord, homeCoord)
        return GeoEngine.relativeAngle(bearingToHome, heading.toDouble())
    }

    private fun computeDistance(location: Location?, homeBase: HomeBase?): Double? {
        if (location == null || homeBase == null) return null

        val currentCoord = GeoEngine.Coordinate(location.latitude, location.longitude)
        val homeCoord = GeoEngine.Coordinate(homeBase.lat, homeBase.lng)

        return GeoEngine.haversine(currentCoord, homeCoord)
    }

    private fun computeWalkTimeString(distance: Double?, location: Location?): String {
        if (distance == null || location == null) return "--"

        val currentCoord = GeoEngine.Coordinate(location.latitude, location.longitude)
        val region = GeoEngine.determineRegion(currentCoord)
        val minutes = GeoEngine.estimateWalkTime(distance, region)

        return when {
            minutes < 1 -> "<1 min"
            minutes == 1 -> "1 min"
            minutes < 60 -> "$minutes min"
            else -> {
                val hours = minutes / 60
                val remainingMinutes = minutes % 60
                if (remainingMinutes == 0) {
                    "$hours hr"
                } else {
                    "$hours hr $remainingMinutes min"
                }
            }
        }
    }

    private fun cardinalDirection(rotation: Double): String {
        val normalized = rotation.mod(360.0)

        return when {
            normalized >= 337.5 || normalized < 22.5 -> "North"
            normalized >= 22.5 && normalized < 67.5 -> "Northeast"
            normalized >= 67.5 && normalized < 112.5 -> "East"
            normalized >= 112.5 && normalized < 157.5 -> "Southeast"
            normalized >= 157.5 && normalized < 202.5 -> "South"
            normalized >= 202.5 && normalized < 247.5 -> "Southwest"
            normalized >= 247.5 && normalized < 292.5 -> "West"
            normalized >= 292.5 && normalized < 337.5 -> "Northwest"
            else -> ""
        }
    }
}

File: ios/MarrakechGuide/MarrakechGuide/Features/HomeBase/HomeBaseViewModel.swift
import SwiftUI
import CoreLocation
import Combine

/// ViewModel for the Home Base compass feature.
///
/// Manages state for:
/// - Home base location setup and retrieval
/// - Current user location tracking
/// - Compass heading for direction to home
/// - Distance and bearing calculations
@MainActor
final class HomeBaseViewModel: ObservableObject {

    // MARK: - Published State

    /// Current home base, nil if not set
    @Published private(set) var homeBase: HomeBase?

    /// Current user location
    @Published private(set) var currentLocation: CLLocation?

    /// Current device heading in degrees
    @Published private(set) var currentHeading: Double?

    /// Confidence level of the heading
    @Published private(set) var headingConfidence: HeadingConfidence = .unavailable

    /// Distance to home base in meters
    @Published private(set) var distanceMeters: Double?

    /// Bearing to home base in degrees
    @Published private(set) var bearingDegrees: Double?

    /// Relative angle for compass arrow (bearing - heading)
    @Published private(set) var arrowRotation: Double = 0

    /// Whether location permission is granted
    @Published private(set) var hasLocationPermission = false

    /// Whether location is being actively tracked
    @Published private(set) var isTracking = false

    /// Error message to display
    @Published var errorMessage: String?

    /// Whether we're loading initial data
    @Published private(set) var isLoading = true

    // MARK: - Dependencies

    private let settingsRepository: UserSettingsRepository
    private let locationService: LocationService
    private let headingService: HeadingService

    // MARK: - Timer for updates

    private var updateTimer: Timer?
    private let updateInterval: TimeInterval = 0.1 // 10Hz

    // MARK: - Initialization

    init(
        settingsRepository: UserSettingsRepository,
        locationService: LocationService = LocationServiceImpl.shared,
        headingService: HeadingService = HeadingServiceImpl.shared
    ) {
        self.settingsRepository = settingsRepository
        self.locationService = locationService
        self.headingService = headingService
    }

    // MARK: - Public Methods

    /// Load initial state
    func load() async {
        isLoading = true
        defer { isLoading = false }

        // Load saved home base
        do {
            homeBase = try await settingsRepository.getHomeBase()
        } catch {
            errorMessage = "Failed to load home base: \(error.localizedDescription)"
        }

        // Check permission status
        hasLocationPermission = locationService.isAuthorized
    }

    /// Request location permission
    func requestPermission() async -> Bool {
        let granted = await locationService.requestPermission()
        hasLocationPermission = granted
        return granted
    }

    /// Save a new home base
    func saveHomeBase(name: String, lat: Double, lng: Double, address: String?) async throws {
        let newHomeBase = HomeBase(name: name, lat: lat, lng: lng, address: address)
        try await settingsRepository.setHomeBase(newHomeBase)
        homeBase = newHomeBase
    }

    /// Set home base to current location
    func setHomeBaseToCurrentLocation(name: String, address: String?) async throws {
        guard let location = locationService.currentLocation else {
            throw HomeBaseError.locationUnavailable
        }

        try await saveHomeBase(
            name: name,
            lat: location.coordinate.latitude,
            lng: location.coordinate.longitude,
            address: address
        )
    }

    /// Start tracking location and heading
    func startTracking() {
        guard homeBase != nil else { return }
        guard hasLocationPermission else { return }

        isTracking = true
        locationService.startUpdates()
        headingService.startUpdates()

        // Start update timer
        updateTimer = Timer.scheduledTimer(withTimeInterval: updateInterval, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.updateCompassState()
            }
        }
    }

    /// Stop tracking
    func stopTracking() {
        isTracking = false
        locationService.stopUpdates()
        headingService.stopUpdates()
        updateTimer?.invalidate()
        updateTimer = nil
    }

    /// Refresh location with high accuracy
    func refreshLocation() async {
        do {
            let location = try await locationService.refreshLocation()
            currentLocation = location
            updateCompassState()
        } catch {
            errorMessage = "Couldn't refresh location: \(error.localizedDescription)"
        }
    }

    // MARK: - Private Methods

    private func updateCompassState() {
        // Update current location
        currentLocation = locationService.currentLocation

        // Update heading
        if let heading = headingService.trueHeadingDegrees {
            currentHeading = heading
        }
        headingConfidence = headingService.headingConfidence

        // Calculate distance and bearing to home base
        guard let home = homeBase,
              let location = currentLocation else {
            distanceMeters = nil
            bearingDegrees = nil
            return
        }

        let homeCoord = GeoEngine.Coordinate(lat: home.lat, lng: home.lng)
        let currentCoord = GeoEngine.Coordinate(
            lat: location.coordinate.latitude,
            lng: location.coordinate.longitude
        )

        // Calculate distance
        distanceMeters = GeoEngine.haversine(from: currentCoord, to: homeCoord)

        // Calculate bearing
        bearingDegrees = GeoEngine.bearing(from: currentCoord, to: homeCoord)

        // Calculate arrow rotation
        if let bearing = bearingDegrees, let heading = currentHeading {
            arrowRotation = GeoEngine.relativeAngle(bearing: bearing, heading: heading)
        }
    }

    // MARK: - Computed Properties

    /// Formatted distance string
    var formattedDistance: String {
        guard let distance = distanceMeters else { return "—" }
        return GeoEngine.formatDistance(meters: distance)
    }

    /// Estimated walk time
    var estimatedWalkTime: String {
        guard let distance = distanceMeters else { return "—" }

        // Determine if in Medina for speed adjustment
        let region: GeoEngine.MarrakechRegion
        if let location = currentLocation {
            let coord = GeoEngine.Coordinate(
                lat: location.coordinate.latitude,
                lng: location.coordinate.longitude
            )
            region = GeoEngine.detectRegion(coordinate: coord)
        } else {
            region = .other
        }

        let minutes = GeoEngine.estimateWalkTime(meters: distance, region: region)

        if minutes < 1 {
            return "< 1 min"
        } else if minutes < 60 {
            return "\(Int(minutes)) min"
        } else {
            let hours = Int(minutes / 60)
            let remainingMins = Int(minutes.truncatingRemainder(dividingBy: 60))
            return "\(hours)h \(remainingMins)m"
        }
    }

    /// Direction description (N, NE, E, etc.)
    var directionDescription: String {
        guard let bearing = bearingDegrees else { return "—" }
        return bearingToCardinal(bearing)
    }

    private func bearingToCardinal(_ bearing: Double) -> String {
        let normalized = bearing.truncatingRemainder(dividingBy: 360)
        let adjusted = normalized < 0 ? normalized + 360 : normalized

        switch adjusted {
        case 337.5..<360, 0..<22.5: return "North"
        case 22.5..<67.5: return "Northeast"
        case 67.5..<112.5: return "East"
        case 112.5..<157.5: return "Southeast"
        case 157.5..<202.5: return "South"
        case 202.5..<247.5: return "Southwest"
        case 247.5..<292.5: return "West"
        case 292.5..<337.5: return "Northwest"
        default: return "—"
        }
    }
}

// MARK: - Errors

enum HomeBaseError: LocalizedError {
    case locationUnavailable
    case permissionDenied

    var errorDescription: String? {
        switch self {
        case .locationUnavailable:
            return "Unable to get current location"
        case .permissionDenied:
            return "Location permission is required"
        }
    }
}
